### PR TITLE
feat: Lua-like stack-based architecture

### PR DESCRIPTION
**Pull request by utkudotdev**
Switch to Lua-like architecture. There are some notes in TODOs about stuff that's bad about this PR so we can discuss. I'm not 100% sure how to best resolve these.

### CODE REVIEW

File: src/opcode.rs [hunk 1]
```diff
@@ -1,20 +1,25 @@
...
+/// Smallest sized integer type that can fit an op code.
+pub type RawOpCode = u8;
+/// Smallest sized integer type that can fit an immediate value.
+pub type Immediate = u16;
+/// Smallest sized integer type that can fit an extended immediate value.
+pub type ExtendedImmediate = u32;
+
 /// Bits for opcode.
-pub const OP_BITS: usize = 7;
+pub const OP_CODE_BITS: usize = 8; // not using ::BITS here because types are just smallest thing that can fit,
```
Comment (ethanuppal) [inline src/opcode.rs]: I prefer code fencing `` all code even in non-doc comments

File: src/opcode.rs [hunk 2]
```diff
@@ -68,26 +79,47 @@ macro_rules! encode {
...
+                result |= ((encoded_final & mask) << offset);
             )*
             let _ = offset;
             result
         }
     };
 }
 
+// TODO: better name?
+/// Traits for types that can be encoded as part of an Op
```
Comment (ethanuppal) [inline src/opcode.rs]:
```rs
/// Argument type for an operation. ???
trait EncodeIntoOp {}
```
btw, shouldn't this return `RawOp`?

File: src/opcode.rs [hunk 3]
```diff
@@ -129,84 +161,80 @@ macro_rules! encode_opcode {
...
+    ($self:expr; $opname:ident as N) => {
+        return Some(encode!(Word;
+            [..OP_CODE_BITS..] = $self.opcode()
+        ));
     };
 }
 
 macro_rules! decoded_opcode {
     ($encoded:expr; $opname:ident as ABC) => {
-        decode!($encoded; $crate::arch::Word;
```
Comment (ethanuppal) [inline src/opcode.rs:169]: This is bad hygiene; in macros, we should always use full paths to types.

File: src/vm.rs [hunk 4]
```diff
@@ -81,29 +87,69 @@ impl VM {
...
         Ok(())
     }
 
-    fn index_from_local(&self, offset: Offset) -> usize {
-        ((self.bp as Offset) + offset) as usize
+    fn ip(&self) -> usize {
+        self.top_frame().ip
+    }
+
+    fn move_ip(&mut self, new_ip: usize) -> Result<(), VMError> {
```
Comment (ethanuppal) [inline src/vm.rs]: can we call this `jump` or smth

Comment (ethanuppal) [reply src/vm.rs]: also should we make a `VMResult` type for convenience

File: src/coding.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                result |= ((encoded_final & mask) << offset);
+            )*
+            let _ = offset;
+            result
+        }
+    };
+}
+
+/// Traits for types that can be encoded as part of an Op
+pub trait Encodable<T> {
```
Comment (ethanuppal) [inline src/coding.rs]: `EncodeInto`? I prefer more declarative names for traits. Do we want these functions in separate traits? If not, `EncodeInto` is a bad name. What about `TruncatedConvert`? That's too long I think.

File: src/coding.rs [hunk 6]
```diff
@@ -27,15 +29,15 @@ macro_rules! encode {
 }
 
 /// Traits for types that can be encoded as part of an Op
```
Comment (ethanuppal) [inline src/coding.rs]: [`Op`] so that it semantically links and looks like code.

File: src/arch.rs [hunk 7]
```diff
@@ -1,19 +1,13 @@
 // Copyright (C) 2024 Ethan Uppal and Utku Melemetci. All rights reserved.
+use core::ops::Range;
```
Comment (ethanuppal) [inline src/arch.rs:3]: nit: Add an empty line above

File: src/coding.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            @(a: u8 = [..8..], b: u8 = [..8..], c: u8 = [..8..]) => {
+                assert_eq!(0, a);
+                assert_eq!(1, b);
+                assert_eq!(2, c);
+            }
+        );
+    }
+
+    #[test]
+    fn decodes_signed_correctly() {
```
Comment (ethanuppal) [inline src/coding.rs:155]: Does sign extension work properly?

File: src/opcode.rs [hunk 9]
```diff
@@ -25,140 +32,101 @@ pub const OFF_BITS: usize = ADDRESS_BITS;
...
+
+    fn decode_from_word(encoded: Word) -> Self {
+        encoded as ExtendedImmediate
+    }
 }
 
 macro_rules! encode_opcode {
     ($self:expr; $opname:ident as ABC) => {
         if let Self::$opname(a, b, c) = $self {
-            return Some(encode!(Word;
```
Comment (ethanuppal) [inline src/opcode.rs:131]: Do you think we should make this `$crate::encode`?

File: src/vm.rs [hunk 10]
```diff
@@ -74,52 +87,184 @@ impl VM {
...
+                        .copy_from_slice(&popped.locals[RETURN_LOCALS]);
                 }
-            };
+
+                self.jump(ra)
+            }
+            Op::Call(offset) => {
+                let as_usize: usize = offset
+                    .try_into()
+                    .expect("illegal offset, too large for platform"); // explodes on microcontrollers
```
Comment (ethanuppal) [inline src/vm.rs:140]: we're only worrying about x64

File: src/vm.rs [hunk 11]
```diff
@@ -74,52 +87,184 @@ impl VM {
...
+    }
+
+    fn current_frame_mut(&mut self) -> &mut StackFrame {
+        self.call_stack.last_mut().expect(
+            "call stack expected to always have one frame while running."
+        )
+    }
+}
+
+fn sign_extend_to<T: Unsigned + PrimInt>(
```
Comment (ethanuppal) [inline src/vm.rs]: Do we really need the entire `num_traits` crate for this?

File: src/vm.rs [hunk 12]
```diff
@@ -74,52 +87,184 @@ impl VM {
...
+            Op::Add(0, 0, 1),
+            Op::Ret(),
+            // func double
+            Op::Mov(1, 0),
+            Op::Call((0 as ExtendedImmediate).wrapping_sub(3)), // call add
+            Op::Ret()
+        ])
+        .expect("invalid program");
+
+        for _ in 0..7 {
```
Comment (ethanuppal) [inline src/vm.rs:261]: Can you factor the program out into a separate variable?

**Review by ethanuppal — CHANGES_REQUESTED** [review]

**Comment by ethanuppal** [comment]
Do you sign extend decoded `ExtendedIntermediate`s?

File: src/opcode.rs (hunk 9 above)
Comment (utkudotdev) [reply src/opcode.rs:131]: Probably

**Comment by ethanuppal** [comment]
Can you run `git rebase --signoff HEAD~N` where `N` is the number of commits you've made up to and including the first unsigned commit?

File: src/coding.rs (hunk 8 above)
Comment (utkudotdev) [reply src/coding.rs:155]: Probably not, but do we want to promise this as part of `decode!`? Currently nothing about `decode!` makes this promise--it is up to the user when implementing `CodeAsWord` to handle this if they feel like it. Our only promise is that we'll give you the `n` bits you asked for by writing `[..n..]` within the `Word` you get in `decode_from_word`.

File: src/vm.rs (hunk 10 above)
Comment (utkudotdev) [reply src/vm.rs:140]: do you think i should replace with `.unwrap()` then?

File: src/vm.rs (hunk 11 above)
Comment (utkudotdev) [reply src/vm.rs]: Nope, fair point. We can probably get this done with a macro.

**Comment by utkudotdev** [comment]
Also, no, decoded `ExtendedImmediate`s are not sign extended right now by default. It is left up to the execution instruction, not the encoding, to do any sign extension if it wants to. I don't know if this is the right design or not, if you have ideas one way or another lmk.

**Comment by utkudotdev** [comment]
Rebased!

File: src/coding.rs (hunk 8 above)
Comment (ethanuppal) [reply src/coding.rs:155]: Good point

File: src/vm.rs (hunk 10 above)
Comment (ethanuppal) [reply src/vm.rs:140]: Actually I realized I don't care

**Review by ethanuppal — APPROVED** [review]

**Pull request merged**